### PR TITLE
Faceting #259

### PR DESCRIPTION
**Pull request by chowington**
Works toward https://github.com/VEuPathDB/web-components/issues/259

### CODE REVIEW

**Comment by dmfalke** [comment]
@chowington I updated the linked issue with some missing details. It may impact the work you've done here.

**Review by bobular — APPROVED** [review]
Hi!
Looks to me like there's a bug in the Storybook Controls UI

In Barplot, for example, I would say there should only be 2 controls: `data` and `props`, but it lists 27.  25 of them are inoperative.  But I guess we shouldn't dwell on this right now.

Looks great otherwise, cheers.

**Comment by bobular** [comment]
> Looks to me like there's a bug in the Storybook Controls UI
> 
> In Barplot, for example, I would say there should only be 2 controls: `data` and `props`, but it lists 27. 25 of them are inoperative. But I guess we shouldn't dwell on this right now.

Maybe the controls are shared per story tsx file?  Or leak between stories in the same file, or something?  Not worth investigating further, just thought I'd make a note of this here.

**Comment by dmfalke** [comment]
I took a very quick look at the code, and I think you're going in the right direction. I haven't tested anything, but hope to do so in the next day or two.

**Comment by chowington** [comment]
@bobular @dmfalke The main tasks for this PR's ticket are mostly done. Do take a look please! There's one outstanding bug and potentially other improvement to make that we can discuss.

**Review by bobular — APPROVED** [review]
This looks great. I tested also by nuking `modalComponentProps` in storybook controls and it works great with the modals disabled.

Am I correct in thinking that if we have N facets, we have N on-screen components that each "contain" data.  We also have N lots of divModalProps.onClick handlers which also contain a copy of the data (in `sharedProps`).  Is that using twice the memory? Or is it just a copy by reference.  And even if it is a deep copy, does it even matter?

Excellent work with the `interactive`/`displayLegend` stuff and I think it is essential to set the modal plot's `title`, so all :+1: from me.

**Comment by bobular** [comment]
Oh there is one small thing.  If you check the Histogram story, you see that the modal plot has a legend with a single item.

In EDA we disable the legend when there's only one data series.  So we will want to be able to control `displayLegend` from the upstream component, rather than have it always `true` as below

```
          const divModalProps = modalComponentProps && {
            onClick: () => {
              setModalPlot(
                <Component
                  {...sharedProps}
                  displayLegend={true}
                  interactive={true}
                  {...modalComponentProps}
                  title={label}
                />
              );
              setModalIsOpen(true);
            },
            style: { cursor: 'pointer' },
            title: 'Click to expand',
          };
```

**Comment by chowington** [comment]
> This looks great. I tested also by nuking `modalComponentProps` in storybook controls and it works great with the modals disabled.
> 
> Am I correct in thinking that if we have N facets, we have N on-screen components that each "contain" data. We also have N lots of divModalProps.onClick handlers which also contain a copy of the data (in `sharedProps`). Is that using twice the memory? Or is it just a copy by reference. And even if it is a deep copy, does it even matter? (thinking scatterplots)
> 
> Excellent work with the `interactive`/`displayLegend` stuff and I think it is essential to set the modal plot's `title`, so all 👍 from me.

Thanks Bob!

I believe JS's object passing behavior is a (slightly-modified) form of pass-by-reference, so theoretically we aren't duplicating the data here. Definitely a valid concern though!

**Comment by chowington** [comment]
> Oh there is one small thing. If you check the Histogram story, you see that the modal plot has a legend with a single item.
> 
> In EDA we disable the legend when there's only one data series. So we will want to be able to control `displayLegend` from the upstream component, rather than have it always `true` as below
> 
> ```
>           const divModalProps = modalComponentProps && {
>             onClick: () => {
>               setModalPlot(
>                 <Component
>                   {...sharedProps}
>                   displayLegend={true}
>                   interactive={true}
>                   {...modalComponentProps}
>                   title={label}
>                 />
>               );
>               setModalIsOpen(true);
>             },
>             style: { cursor: 'pointer' },
>             title: 'Click to expand',
>           };
> ```

This should already allow that! The current logic allows you to override `displayLegend={true}` by passing your own `displayLegend` value in `modalComponentProps`, as `{...modalComponentProps}` overwrites the previous props 👍

**Comment by bobular** [comment]
Oh yeah, sorry I didn't spot that!

**Comment by chowington** [comment]
> @bobular @dmfalke The main tasks for this PR's ticket are mostly done. Do take a look please! There's one outstanding bug and potentially other improvement to make that we can discuss.

Sorry, I was a little vague late last night! More details:

- Bug: Clicking directly on the pie chart portion of a pie facet doesn't trigger the click handler that opens the modal. You have to click outside the pie itself. I'm not sure why exactly this is happening yet; if anyone has an idea, let me know!
- Comment: I'll probably double check the layout prop defaults to make sure they are reasonable
- Question: Should we have defaults for the modal component layout props as well?
- Question: How should we handle the mosaic contingency table? Should there be a modal for those tables as well? I didn't test that at all and there isn't a story for the cont table, faceted or otherwise

**Comment by chowington** [comment]
@dmfalke @bobular Any comments on the above?

**Comment by bobular** [comment]
Thanks @chowington - having a look now.

Unrelated to your questions, I saw this in the console, FYI:

> error: "Warning: react-modal: App element is not defined. Please use `Modal.setAppElement(el)` or set `appElement={el}`. This is needed so screen readers don't see main content when modal is opened. It is not recommended, but you can opt-out by setting `ariaHideApp={false}`."

**Comment by dmfalke** [comment]
> > @bobular @dmfalke The main tasks for this PR's ticket are mostly done. Do take a look please! There's one outstanding bug and potentially other improvement to make that we can discuss.
> 
> Sorry, I was a little vague late last night! More details:
> 
>     * [x]  Bug (Fixed): Clicking directly on the pie chart portion of a pie facet doesn't trigger the click handler that opens the modal. You have to click outside the pie itself. I'm not sure why exactly this is happening yet; if anyone has an idea, let me know!
> 
>     * [ ]  Comment: I'll probably double check the layout prop defaults to make sure they are reasonable
> 
>     * [ ]  Question: Should we have defaults for the modal component layout props as well?
I think it makes sense to have defaults.

> 
>     * [ ]  Question: How should we handle the mosaic contingency table? Should there be a modal for those tables as well? I didn't test that at all and there isn't a story for the cont table, faceted or otherwise
My initial thought is to have the same set of tabs we include for an unfaceted mosaic plot.

File: src/plots/facetedPlots/FacetedPiePlot.tsx [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+type FacetedPiePlotProps = Omit<
+  FacetedPlotProps<PiePlotData, PiePlotProps>,
+  'component'
+>;
+
+const FacetedPiePlot = (facetedPiePlotProps: FacetedPiePlotProps) => {
+  const onPlotlyRender = useCallback((_, graphDiv) => {
+    // Replace each slice DOM node with a copy of itself.
+    // This removes the existing click event handler, which
+    // otherwise blocks the click handler that opens the modal
```
Comment (bobular) [inline src/plots/facetedPlots/FacetedPiePlot.tsx:31]: Thanks for the comment - was looking at an older commit and was going to ask for one!

**Review by bobular — APPROVED** [review]
Code looks good, works great.

It would be nice to have the mosaic modal have the table tabs too, but isn't this going to be a nightmare to implement?  Instead of passing `modalComponentProps` to `FacetedMosaicPlot` we'll have to pass an actual component.  Maybe that's an OK hack?

**Comment by dmfalke** [comment]
> Code looks good, works great.
> 
> It would be nice to have the mosaic modal have the table tabs too, but isn't this going to be a nightmare to implement? Instead of passing `modalComponentProps` to `FacetedMosaicPlot` we'll have to pass an actual component. Maybe that's an OK hack?

You're right, I didn't think about that. Let's just stick with the plot for now and see what others think.

**Review by dmfalke — APPROVED** [review]
I think this all looks great. One suggestion is to export the default styling and spacing options so that consumers can use them if passing in custom styling or spacing.

**Comment by chowington** [comment]
> Thanks @chowington - having a look now.
> 
> Unrelated to your questions, I saw this in the console, FYI:
> 
> > error: "Warning: react-modal: App element is not defined. Please use `Modal.setAppElement(el)` or set `appElement={el}`. This is needed so screen readers don't see main content when modal is opened. It is not recommended, but you can opt-out by setting `ariaHideApp={false}`."

Yeah, I saw that. The modal comes from `core-components`, and Mike said he'll address that console warning at some point 👍

**Comment by chowington** [comment]
> > * [ ]  Question: Should we have defaults for the modal component layout props as well?
> 
> I think it makes sense to have defaults.

I thought about and worked on this a little bit and I think the resulting API would be a bit odd. Currently `modalComponentProps` existing is the way we activate the click-to-expand feature. If we provide sensible defaults, it would be odd for the user to still have to provide `modalComponentProps` in order to activate the click-to-expand feature.

Instead I attempted to create a better API by providing a `clickToExpand?: boolean` prop that controls the expand feature's activation, however after changing the logic appropriately, there was a complicated TS error that cropped up involving the `P` generic type. I think I understood it, but I didn't see a way to get around it.

So what all this means is that I decided to stick with what we have currently, including not giving default `modalComponentsProps`. How does this sound @dmfalke?

**Comment by chowington** [comment]
> Code looks good, works great.
> 
> It would be nice to have the mosaic modal have the table tabs too, but isn't this going to be a nightmare to implement? Instead of passing `modalComponentProps` to `FacetedMosaicPlot` we'll have to pass an actual component. Maybe that's an OK hack?

Alternatively, we could have the click-to-expand feature for the contingency tables as well, separate from the faceted mosaics. Would that be useful? I forgot what the contingency tables look like in faceted mode.

I am itching to get this merged though, because if another PR gets merge beforehand it could be a nasty merge conflict :)

**Comment by chowington** [comment]
I'm getting ready to wrap this up on the `web-components` side. Just wanted to get one more review from @dmfalke on this and I'll merge: https://github.com/VEuPathDB/web-components/pull/272/commits/ac3926d128d9cfd860506ccbb8b5e25f00e9da4a

**Review by dmfalke — APPROVED** [review]
Looks good!

File: src/plots/FacetedPlot.tsx [hunk 2]
```diff
@@ -6,36 +6,51 @@ import React, {
...
+   */
+  modalComponentProps?: P;
   // custom legend prop
   checkedLegendItems?: string[];
 }
 
+export interface FacetedPlotPropsWithRef<D, P extends PlotProps<D>>
+  extends FacetedPlotProps<D, P> {
+  facetedPlotRef?: Ref<FacetedPlotRef>;
+}
```
Comment (dmfalke) [inline src/plots/FacetedPlot.tsx:38]: :+1:

**Comment by chowington** [comment]
Thanks Dave!

**Pull request merged**